cpp/test/sg/decisiontree_batchedlevel_unittest.cu

```diff
@@ -0,0 +1,227 @@
+  }
+
+  DecisionTreeParams params;
+
+  std::unique_ptr<raft::handle_t> raft_handle;
+  std::shared_ptr<TemporaryMemory<DataT, LabelT>> tempmem;
+
+  DataT* h_quantiles;
+  Traits::InputT input;
+
+  NodeT* curr_nodes;
+  NodeT* new_nodes;
+  IdxT* n_new_nodes;
+  IdxT* n_new_leaves;
+  IdxT* new_depth;
+  Traits::SplitT* splits;
+
+  DataT* data;
+  DataT* labels;
+  IdxT* row_ids;
+  IdxT* col_ids;
+};
+
+TEST_F(NodeSplitKernelUnitTest, Quantiles) {
+  /* Ensure that quantiles are computed correctly */
+  std::vector<DataT> expected_quantiles[]{{-2.0f, -1.0f, 0.0f, 2.0f},
+                                          {0.0f, 1.0f, 3.0f}};
+  for (int col = 0; col < n_col; col++) {
+    std::vector<DataT> col_quantile(n_bins);
+    std::copy(h_quantiles + n_bins * col, h_quantiles + n_bins * (col + 1),
+              col_quantile.begin());
+    auto last = std::unique(col_quantile.begin(), col_quantile.end());
+    col_quantile.erase(last, col_quantile.end());
+    EXPECT_EQ(col_quantile, expected_quantiles[col]);
+  }
+}
+
+TEST_P(NodeSplitKernelUnitTest, MinSamplesSplitLeaf) {
+  auto test_params = GetParam();
+
+  Builder<Traits> builder;
+  auto smemSize = Traits::nodeSplitSmemSize(builder);
+
+  IdxT h_n_total_nodes = 3;  // total number of nodes created so far
+  IdxT h_n_new_nodes;        // number of nodes created in this round
+  IdxT batchSize = 2;
+  std::vector<NodeT> h_nodes{
+    /* {
+     *   SparseTreeNode{
+     *     prediction, colid, quesval, best_metric_val, left_child_id },
+     *   }, start, count, depth
+     * } */
+    {{1.40f, 0, -0.5f, 5.606667f, 1}, 0, 5, 0},
+    {{-1.50f, IdxT(-1), DataT(0), DataT(0), NodeT::Leaf}, 0, 2, 1},
+    {{3.333333f, IdxT(-1), DataT(0), DataT(0), NodeT::Leaf}, 1, 3, 1},
+  };
+  raft::update_device(curr_nodes, h_nodes.data() + 1, batchSize, 0);
+  CUDA_CHECK(cudaMemsetAsync(n_new_nodes, 0, sizeof(IdxT), 0));
+  CUDA_CHECK(cudaMemsetAsync(n_new_leaves, 0, sizeof(IdxT), 0));
+  CUDA_CHECK(cudaMemsetAsync(new_depth, 0, sizeof(IdxT), 0));
+  initSplit<DataT, IdxT, Traits::TPB_DEFAULT>(splits, batchSize, 0);
+
+  /* { quesval, colid, best_metric_val, nLeft } */
+  std::vector<Traits::SplitT> h_splits{{-1.5f, 0, 0.25f, 1},
+                                       {2.0f, 1, 3.555556f, 2}};
+  raft::update_device(splits, h_splits.data(), 2, 0);
+
+  nodeSplitKernel<DataT, LabelT, IdxT, Traits::DevTraits, Traits::TPB_SPLIT>
+    <<<batchSize, Traits::TPB_SPLIT, smemSize, 0>>>(
+      params.max_depth, test_params.min_samples_leaf,
+      test_params.min_samples_split, params.max_leaves,
+      params.min_impurity_decrease, input, curr_nodes, new_nodes, n_new_nodes,
+      splits, n_new_leaves, h_n_total_nodes, new_depth);
+  CUDA_CHECK(cudaGetLastError());
+  raft::update_host(&h_n_new_nodes, n_new_nodes, 1, 0);
+  CUDA_CHECK(cudaStreamSynchronize(0));
+  h_n_total_nodes += h_n_new_nodes;
+  EXPECT_EQ(h_n_total_nodes, test_params.expected_n_total_nodes);
```

Comment:
The node splitting test seems useful to make sure we're splitting right... I was thinking you'd actually test the impurity calculation too (against a golden value computed offline).. is that test possible as well?

Reply:
Yes, I will do that in another PR.

Reply:
works for me